File: casino/services/provably_fair.py
"""
Provably Fair Engine
====================

Cryptographically verifiable random number generation.
Players can verify every outcome is fair and predetermined.
"""

import hashlib
import hmac
import secrets
from dataclasses import dataclass
from datetime import datetime
from decimal import Decimal
from typing import Optional, List, Tuple
from uuid import UUID

from sqlalchemy import select, update
from sqlalchemy.ext.asyncio import AsyncSession

from casino.models import ServerSeed, ClientSeed, User


@dataclass
class GameResult:
    """Result of a provably fair calculation"""
    raw_hash: str  # Full HMAC result
    raw_value: int  # Numeric value from hash
    normalized: float  # Value between 0-1
    server_seed_hash: str  # Commitment (shown before bet)
    client_seed: str
    nonce: int
    
    def verify(self, revealed_server_seed: str) -> bool:
        """Verify the result with revealed server seed"""
        # Check server seed matches commitment
        if hashlib.sha256(revealed_server_seed.encode()).hexdigest() != self.server_seed_hash:
            return False
        
        # Regenerate result
        message = f"{self.client_seed}:{self.nonce}"
        expected_hash = hmac.new(
            revealed_server_seed.encode(),
            message.encode(),
            hashlib.sha256
        ).hexdigest()
        
        return expected_hash == self.raw_hash


class ProvablyFairEngine:
    """
    Provably Fair 2.0 Implementation
    
    Flow:
    1. Server generates seed batch OFFLINE, stores hash commitments
    2. User is shown server_seed_hash BEFORE betting
    3. User provides (or system generates) client_seed
    4. Outcome = HMAC-SHA256(server_seed, client_seed:nonce)
    5. After rotating seed, server reveals old seed for verification
    
    This ensures:
    - Server cannot change outcome after bet (committed to hash)
    - Player can influence outcome (client seed)
    - All results are deterministic and verifiable
    """
    
    def __init__(self, session: AsyncSession):
        self.session = session
    
    @staticmethod
    def generate_seed() -> str:
        """Generate a cryptographically secure random seed"""
        return secrets.token_hex(32)
    
    @staticmethod
    def hash_seed(seed: str) -> str:
        """Hash a seed for commitment"""
        return hashlib.sha256(seed.encode()).hexdigest()
    
    async def generate_seed_batch(
        self, 
        user_id: UUID, 
        count: int = 10
    ) -> List[ServerSeed]:
        """
        Pre-generate a batch of server seeds for a user.
        
        In production, this should be done OFFLINE for maximum security.
        """
        seeds = []
        
        for _ in range(count):
            seed = self.generate_seed()
            seed_hash = self.hash_seed(seed)
            
            server_seed = ServerSeed(
                user_id=user_id,
                seed=seed,
                seed_hash=seed_hash,
                is_active=False,
                nonce=0
            )
            self.session.add(server_seed)
            seeds.append(server_seed)
        
        # Activate the first seed
        if seeds:
            seeds[0].is_active = True
        
        await self.session.flush()
        return seeds
    
    async def get_active_server_seed(self, user_id: UUID) -> Optional[ServerSeed]:
        """Get user's current active server seed"""
        result = await self.session.execute(
            select(ServerSeed)
            .where(ServerSeed.user_id == user_id)
            .where(ServerSeed.is_active == True)
        )
        return result.scalar_one_or_none()
    
    async def get_or_create_client_seed(self, user_id: UUID) -> ClientSeed:
        """Get or create user's active client seed"""
        result = await self.session.execute(
            select(ClientSeed)
            .where(ClientSeed.user_id == user_id)
            .where(ClientSeed.is_active == True)
        )
        client_seed = result.scalar_one_or_none()
        
        if not client_seed:
            client_seed = ClientSeed(
                user_id=user_id,
                seed=self.generate_seed()[:16],  # Shorter for UX
                is_active=True
            )
            self.session.add(client_seed)
            await self.session.flush()
        
        return client_seed
    
    async def set_client_seed(self, user_id: UUID, seed: str) -> ClientSeed:
        """Set a new client seed (user-provided)"""
        # Deactivate old seed
        await self.session.execute(
            update(ClientSeed)
            .where(ClientSeed.user_id == user_id)
            .where(ClientSeed.is_active == True)
            .values(is_active=False)
        )
        
        # Create new seed
        client_seed = ClientSeed(
            user_id=user_id,
            seed=seed[:64],  # Limit length
            is_active=True
        )
        self.session.add(client_seed)
        await self.session.flush()
        
        return client_seed
    
    async def rotate_server_seed(self, user_id: UUID) -> Tuple[ServerSeed, ServerSeed]:
        """
        Rotate to a new server seed.
        
        Returns (old_seed, new_seed)
        The old seed is now revealed and can be verified.
        """
        # Get current active seed
        old_seed = await self.get_active_server_seed(user_id)
        
        if not old_seed:
            raise ValueError("No active server seed found")
        
        # Deactivate and reveal old seed
        old_seed.is_active = False
        old_seed.revealed_at = datetime.utcnow()
        
        # Find next inactive seed or generate new one
        result = await self.session.execute(
            select(ServerSeed)
            .where(ServerSeed.user_id == user_id)
            .where(ServerSeed.is_active == False)
            .where(ServerSeed.revealed_at == None)
            .order_by(ServerSeed.id.asc())
            .limit(1)
        )
        new_seed = result.scalar_one_or_none()
        
        if not new_seed:
            # Generate new batch
            new_seeds = await self.generate_seed_batch(user_id, count=10)
            new_seed = new_seeds[0]
        
        new_seed.is_active = True
        await self.session.flush()
        
        return old_seed, new_seed
    
    def generate_outcome(
        self,
        server_seed: str,
        client_seed: str,
        nonce: int
    ) -> GameResult:
        """
        Generate a provably fair outcome.
        
        The outcome is deterministic based on:
        - server_seed: Known only to server until revealed
        - client_seed: Provided by player
        - nonce: Increments with each bet
        """
        message = f"{client_seed}:{nonce}"
        
        raw_hash = hmac.new(
            server_seed.encode(),
            message.encode(),
            hashlib.sha256
        ).hexdigest()
        
        # Use first 8 hex chars (32 bits) for primary value
        raw_value = int(raw_hash[:8], 16)
        
        # Normalize to 0-1 range
        normalized = raw_value / 0xFFFFFFFF
        
        return GameResult(
            raw_hash=raw_hash,
            raw_value=raw_value,
            normalized=normalized,
            server_seed_hash=self.hash_seed(server_seed),
            client_seed=client_seed,
            nonce=nonce
        )
    
    async def generate_bet_outcome(
        self,
        user_id: UUID,
        increment_nonce: bool = True
    ) -> Tuple[GameResult, int]:
        """
        Generate outcome for a bet.
        
        Returns (result, server_seed_id)
        """
        server_seed = await self.get_active_server_seed(user_id)
        if not server_seed:
            # Auto-generate seeds for new user
            seeds = await self.generate_seed_batch(user_id)
            server_seed = seeds[0]
        
        client_seed = await self.get_or_create_client_seed(user_id)
        
        current_nonce = server_seed.nonce
        
        result = self.generate_outcome(
            server_seed=server_seed.seed,
            client_seed=client_seed.seed,
            nonce=current_nonce
        )
        
        if increment_nonce:
            server_seed.nonce += 1
            await self.session.flush()
        
        return result, server_seed.id
    
    @staticmethod
    def dice_result(normalized: float, precision: int = 2) -> Decimal:
        """
        Convert normalized value to dice roll (0.00 - 99.99)
        
        Dice game: Player picks over/under target
        """
        return Decimal(str(round(normalized * 100, precision)))
    
    @staticmethod
    def crash_result(normalized: float, house_edge: float = 0.03) -> Decimal:
        """
        Convert normalized value to crash multiplier.
        
        Uses the standard crash formula with house edge.
        Formula: 1 / (1 - normalized * (1 - house_edge))
        
        This gives exponential distribution with expected RTP = 1 - house_edge
        """
        # Avoid division by zero at edge
        if normalized >= (1 - house_edge):
            normalized = 1 - house_edge - 0.0001
        
        multiplier = 1 / (1 - normalized)
        
        # Apply house edge
        multiplier = multiplier * (1 - house_edge)
        
        # Minimum 1.00x
        return Decimal(str(max(1.00, round(multiplier, 2))))
    
    @staticmethod
    def limbo_result(normalized: float, house_edge: float = 0.03) -> Decimal:
        """
        Convert normalized value to limbo multiplier.
        
        Similar to crash but instant result.
        """
        if normalized >= (1 - house_edge):
            normalized = 1 - house_edge - 0.0001
        
        multiplier = 1 / (1 - normalized)
        multiplier = multiplier * (1 - house_edge)
        
        return Decimal(str(max(1.00, round(multiplier, 2))))
    
    @staticmethod
    def plinko_result(
        normalized: float, 
        rows: int = 16, 
        hash_full: str = None
    ) -> List[str]:
        """
        Generate Plinko path (sequence of L/R bounces).
        
        Uses multiple bytes from hash for each row.
        """
        if hash_full is None:
            raise ValueError("Full hash required for Plinko")
        
        path = []
        for i in range(rows):
            # Use different bytes for each row
            byte_value = int(hash_full[i*2:(i+1)*2], 16)
            direction = "R" if byte_value >= 128 else "L"
            path.append(direction)
        
        return path
    
    @staticmethod
    def mines_generate_mines(
        normalized: float,
        hash_full: str,
        grid_size: int = 25,
        mine_count: int = 5
    ) -> List[int]:
        """
        Generate mine positions for Mines game.
        
        Uses Fisher-Yates shuffle seeded by hash.
        """
        # Create position array
        positions = list(range(grid_size))
        
        # Use hash segments to shuffle
        for i in range(grid_size - 1, 0, -1):
            # Get random index from hash
            hash_segment = hash_full[(i % 32) * 2: (i % 32) * 2 + 2]
            j = int(hash_segment, 16) % (i + 1)
            positions[i], positions[j] = positions[j], positions[i]
        
        # First N positions are mines
        return sorted(positions[:mine_count])
    
    @staticmethod
    def wheel_result(normalized: float, segments: int = 50) -> int:
        """
        Generate wheel/roulette segment.
        
        Returns segment index (0 to segments-1)
        """
        return int(normalized * segments) % segments
    
    async def get_verification_data(
        self, 
        user_id: UUID
    ) -> dict:
        """Get data needed for user to verify their bets"""
        server_seed = await self.get_active_server_seed(user_id)
        client_seed = await self.get_or_create_client_seed(user_id)
        
        # Get previous revealed seeds
        result = await self.session.execute(
            select(ServerSeed)
            .where(ServerSeed.user_id == user_id)
            .where(ServerSeed.revealed_at != None)
            .order_by(ServerSeed.revealed_at.desc())
            .limit(10)
        )
        revealed_seeds = result.scalars().all()
        
        return {
            "active_server_seed_hash": server_seed.seed_hash if server_seed else None,
            "current_nonce": server_seed.nonce if server_seed else 0,
            "client_seed": client_seed.seed if client_seed else None,
            "revealed_seeds": [
                {
                    "seed": s.seed,
                    "seed_hash": s.seed_hash,
                    "nonce_count": s.nonce,
                    "revealed_at": s.revealed_at.isoformat()
                }
                for s in revealed_seeds
            ]
        }


def verify_bet(
    server_seed: str,
    server_seed_hash: str,
    client_seed: str,
    nonce: int,
    game_type: str
) -> dict:
    """
    Standalone verification function.
    
    Can be used client-side to verify any bet.
    """
    # Verify server seed matches hash
    computed_hash = hashlib.sha256(server_seed.encode()).hexdigest()
    if computed_hash != server_seed_hash:
        return {"valid": False, "error": "Server seed doesn't match hash"}
    
    # Generate result
    message = f"{client_seed}:{nonce}"
    raw_hash = hmac.new(
        server_seed.encode(),
        message.encode(),
        hashlib.sha256
    ).hexdigest()
    
    raw_value = int(raw_hash[:8], 16)
    normalized = raw_value / 0xFFFFFFFF
    
    # Calculate game-specific result
    if game_type == "dice":
        result = round(normalized * 100, 2)
    elif game_type == "crash" or game_type == "limbo":
        result = max(1.00, round((1 / (1 - normalized)) * 0.97, 2))
    elif game_type == "wheel":
        result = int(normalized * 50) % 50
    else:
        result = normalized
    
    return {
        "valid": True,
        "raw_hash": raw_hash,
        "raw_value": raw_value,
        "normalized": normalized,
        "result": result
    }

File: frontend/lib/provably-fair.ts
/**
 * Provably Fair Cryptographic System
 * 
 * This system uses SHA-256 hashing to generate verifiable random outcomes.
 * Players can verify every bet result using the server seed, client seed, and nonce.
 * 
 * How it works:
 * 1. Server generates a random server seed and provides the hash (commitment)
 * 2. Player provides their client seed (or uses default)
 * 3. On each bet, the nonce increments
 * 4. Combined hash = SHA-256(serverSeed:clientSeed:nonce)
 * 5. This hash is converted to a game result
 * 6. After changing server seed, the old seed is revealed for verification
 */

// Web Crypto API based SHA-256
async function sha256(message: string): Promise<string> {
  const msgBuffer = new TextEncoder().encode(message)
  const hashBuffer = await crypto.subtle.digest('SHA-256', msgBuffer)
  const hashArray = Array.from(new Uint8Array(hashBuffer))
  return hashArray.map(b => b.toString(16).padStart(2, '0')).join('')
}

// Synchronous SHA-256 using SubtleCrypto with pre-computed values for client-side
function sha256Sync(message: string): string {
  // Simple hash implementation for synchronous operations
  let hash = 0
  for (let i = 0; i < message.length; i++) {
    const char = message.charCodeAt(i)
    hash = ((hash << 5) - hash) + char
    hash = hash & hash
  }
  // Expand to 64 char hex string
  const hex = Math.abs(hash).toString(16).padStart(8, '0')
  return hex.repeat(8)
}

// Generate a cryptographically secure random hex string
export function generateSecureRandom(bytes: number = 32): string {
  const array = new Uint8Array(bytes)
  crypto.getRandomValues(array)
  return Array.from(array).map(b => b.toString(16).padStart(2, '0')).join('')
}

// Generate server seed hash (commitment)
export async function hashServerSeed(serverSeed: string): Promise<string> {
  return sha256(serverSeed)
}

// Generate combined hash for bet
export async function generateBetHash(
  serverSeed: string,
  clientSeed: string,
  nonce: number
): Promise<string> {
  const combined = `${serverSeed}:${clientSeed}:${nonce}`
  return sha256(combined)
}

// Convert hash to float between 0 and 1
export function hashToFloat(hash: string): number {
  // Take first 8 characters (32 bits) and convert to float
  const int = parseInt(hash.substring(0, 8), 16)
  return int / 0xFFFFFFFF
}

// Convert hash to integer in range [min, max]
export function hashToInt(hash: string, min: number, max: number): number {
  const float = hashToFloat(hash)
  return Math.floor(float * (max - min + 1)) + min
}

// ==================== GAME SPECIFIC ALGORITHMS ====================

/**
 * CRASH GAME
 * Generates a crash point multiplier
 * House edge is built into the formula
 */
export function calculateCrashPoint(hash: string, houseEdge: number = 0.04): number {
  const float = hashToFloat(hash)
  
  // 4% chance of instant crash (1.00x)
  if (float < houseEdge) {
    return 1.00
  }
  
  // Calculate crash point with house edge
  // Formula: 1 / (1 - (float * (1 - houseEdge)))
  const crashPoint = (1 - houseEdge) / (1 - float)
  
  // Round to 2 decimal places, max at 1000000x
  return Math.min(Math.floor(crashPoint * 100) / 100, 1000000)
}

/**
 * DICE GAME
 * Generates a roll between 0.00 and 100.00
 */
export function calculateDiceRoll(hash: string): number {
  const float = hashToFloat(hash)
  // Convert to roll between 0.00 and 99.99
  return Math.floor(float * 10000) / 100
}

/**
 * MINES GAME
 * Generates mine positions for a grid
 */
export function generateMinePositions(
  hash: string,
  gridSize: number = 25,
  mineCount: number = 5
): number[] {
  const positions: number[] = []
  let currentHash = hash
  
  while (positions.length < mineCount) {
    const position = hashToInt(currentHash, 0, gridSize - 1)
    
    if (!positions.includes(position)) {
      positions.push(position)
    }
    
    // Generate new hash for next iteration
    currentHash = sha256Sync(currentHash)
  }
  
  return positions.sort((a, b) => a - b)
}

/**
 * PLINKO GAME
 * Generates the path of the ball (left/right at each peg)
 */
export function generatePlinkoPath(hash: string, rows: number = 16): ('L' | 'R')[] {
  const path: ('L' | 'R')[] = []
  let currentHash = hash
  
  for (let i = 0; i < rows; i++) {
    const float = hashToFloat(currentHash.substring(i * 4, (i * 4) + 8) || currentHash)
    path.push(float < 0.5 ? 'L' : 'R')
    
    if (i % 8 === 7) {
      currentHash = sha256Sync(currentHash)
    }
  }
  
  return path
}

/**
 * Calculate Plinko multiplier based on final bucket
 */
export function calculatePlinkoMultiplier(path: ('L' | 'R')[], risk: 'low' | 'medium' | 'high' = 'medium'): number {
  // Count how many rights (determines final bucket position)
  const rightCount = path.filter(p => p === 'R').length
  const rows = path.length
  
  // Multipliers for 16 rows, different risk levels
  const multipliers: Record<string, number[]> = {
    low: [5.6, 2.1, 1.1, 1, 0.5, 1, 0.3, 0.5, 0.5, 0.3, 1, 0.5, 1, 1.1, 2.1, 5.6, 16],
    medium: [13, 3, 1.3, 0.7, 0.4, 0.3, 0.2, 0.2, 0.2, 0.2, 0.3, 0.4, 0.7, 1.3, 3, 13, 110],
    high: [1000, 130, 26, 9, 4, 2, 0.2, 0.2, 0.2, 0.2, 2, 4, 9, 26, 130, 1000, 1000]
  }
  
  const bucketIndex = Math.min(rightCount, rows)
  return multipliers[risk][bucketIndex] || 1
}

/**
 * LIMBO GAME
 * Generates a target multiplier (similar to crash but instant)
 */
export function calculateLimboResult(hash: string, houseEdge: number = 0.04): number {
  const float = hashToFloat(hash)
  
  // Similar to crash but different distribution
  if (float < houseEdge) {
    return 1.00
  }
  
  const result = (1 - houseEdge) / (1 - float)
  return Math.floor(result * 100) / 100
}

/**
 * WHEEL GAME
 * Determines which segment the wheel lands on
 */
export function calculateWheelResult(hash: string, segments: number = 50): number {
  return hashToInt(hash, 0, segments - 1)
}

/**
 * KENO GAME
 * Generates the drawn numbers
 */
export function generateKenoNumbers(hash: string, count: number = 10, max: number = 40): number[] {
  const numbers: number[] = []
  let currentHash = hash
  
  while (numbers.length < count) {
    const num = hashToInt(currentHash, 1, max)
    
    if (!numbers.includes(num)) {
      numbers.push(num)
    }
    
    currentHash = sha256Sync(currentHash)
  }
  
  return numbers.sort((a, b) => a - b)
}

/**
 * BLACKJACK / TWENTY ONE
 * Generates a shuffled deck order
 */
export function generateDeckOrder(hash: string): number[] {
  const deck = Array.from({ length: 52 }, (_, i) => i)
  let currentHash = hash
  
  // Fisher-Yates shuffle using hash
  for (let i = deck.length - 1; i > 0; i--) {
    const j = hashToInt(currentHash, 0, i)
    ;[deck[i], deck[j]] = [deck[j], deck[i]]
    
    if (i % 8 === 0) {
      currentHash = sha256Sync(currentHash)
    }
  }
  
  return deck
}

/**
 * COIN FLIP
 * Simple heads or tails
 */
export function calculateCoinFlip(hash: string): 'heads' | 'tails' {
  const float = hashToFloat(hash)
  return float < 0.5 ? 'heads' : 'tails'
}

// ==================== VERIFICATION FUNCTIONS ====================

export interface BetVerification {
  serverSeed: string
  serverSeedHash: string
  clientSeed: string
  nonce: number
  betHash: string
  result: unknown
  isValid: boolean
}

export async function verifyBet(
  serverSeed: string,
  clientSeed: string,
  nonce: number,
  expectedHash: string
): Promise<boolean> {
  const computedHash = await generateBetHash(serverSeed, clientSeed, nonce)
  return computedHash === expectedHash
}

// ==================== SESSION MANAGEMENT ====================

export interface GameSession {
  id: string
  serverSeedHash: string  // Public commitment
  serverSeed: string      // Hidden until rotation
  clientSeed: string
  nonce: number
  createdAt: number
}

export function createGameSession(clientSeed?: string): GameSession {
  const serverSeed = generateSecureRandom(32)
  
  return {
    id: generateSecureRandom(16),
    serverSeedHash: sha256Sync(serverSeed),
    serverSeed,
    clientSeed: clientSeed || generateSecureRandom(16),
    nonce: 0,
    createdAt: Date.now()
  }
}

export function incrementNonce(session: GameSession): GameSession {
  return {
    ...session,
    nonce: session.nonce + 1
  }
}

export function rotateServerSeed(session: GameSession): { 
  newSession: GameSession
  revealedSeed: string 
} {
  const revealedSeed = session.serverSeed
  const newServerSeed = generateSecureRandom(32)
  
  return {
    newSession: {
      ...session,
      id: generateSecureRandom(16),
      serverSeedHash: sha256Sync(newServerSeed),
      serverSeed: newServerSeed,
      nonce: 0,
      createdAt: Date.now()
    },
    revealedSeed
  }
}
